File: server/scraper.ts
import axios from "axios";
import * as cheerio from "cheerio";
import type { InsertOpinion } from "@shared/schema";

const CAFC_BASE_URL = "https://www.cafc.uscourts.gov";
const OPINIONS_URL = `${CAFC_BASE_URL}/home/case-information/opinions-orders/`;

export interface ScrapedOpinion {
  caseName: string;
  appealNo: string;
  releaseDate: string;
  status: string;
  origin: string;
  documentType: string;
  pdfUrl: string;
  summary: string;
}

export async function scrapeCAFCOpinions(maxPages: number = 1): Promise<ScrapedOpinion[]> {
  const opinions: ScrapedOpinion[] = [];
  
  try {
    console.log(`Fetching CAFC opinions from ${OPINIONS_URL}...`);
    const response = await axios.get(OPINIONS_URL, {
      timeout: 30000,
      headers: {
        'User-Agent': 'Mozilla/5.0 (compatible; CAFC-Copilot/1.0)',
      }
    });
    
    const $ = cheerio.load(response.data);
    
    // The CAFC page uses a table structure for opinions
    // Table columns: Release Date, Appeal Number, Origin, Document Type, Case Name (with link), Status, PDF Path
    $('tr[id^="table_1_row_"]').each((_, row) => {
      const $row = $(row);
      const columns = $row.find('td');
      
      if (columns.length >= 6) {
        const releaseDate = $(columns[0]).text().trim();
        const appealNo = $(columns[1]).text().trim();
        const origin = $(columns[2]).text().trim();
        const documentType = $(columns[3]).text().trim();
        const status = $(columns[5]).text().trim();
        
        // Find the PDF link in the case name column (index 4)
        const $link = $(columns[4]).find('a');
        const caseNameRaw = $link.text().trim();
        const pdfPath = $link.attr('href');
        
        // Clean up case name (remove [OPINION] or [ORDER] suffix)
        const caseName = caseNameRaw.replace(/\s*\[(OPINION|ORDER)\]\s*$/i, '').trim();
        
        if (pdfPath && caseName && status === "Precedential" && documentType === "OPINION") {
          const pdfUrl = pdfPath.startsWith('http') 
            ? pdfPath 
            : `${CAFC_BASE_URL}${pdfPath.startsWith('/') ? pdfPath : '/' + pdfPath}`;
          
          opinions.push({
            caseName,
            appealNo,
            releaseDate,
            status,
            origin,
            documentType,
            pdfUrl,
            summary: `${status} ${documentType} from ${origin}`,
          });
        }
      }
    });
    
    console.log(`Scraped ${opinions.length} precedential opinions from CAFC`);
    
  } catch (error) {
    console.error('Error scraping CAFC opinions:', error);
    throw new Error(`Failed to scrape CAFC opinions: ${error instanceof Error ? error.message : 'Unknown error'}`);
  }
  
  return opinions;
}

export function convertToInsertOpinion(scraped: ScrapedOpinion): InsertOpinion {
  return {
    caseName: scraped.caseName,
    appealNo: scraped.appealNo,
    releaseDate: scraped.releaseDate,
    status: scraped.status,
    origin: scraped.origin,
    documentType: scraped.documentType,
    pdfUrl: scraped.pdfUrl,
    summary: scraped.summary,
    isIngested: false,
    pdfText: null,
  };
}

File: backend/scraper.py
import httpx
from bs4 import BeautifulSoup
from typing import List, Dict
import re

CAFC_URL = "https://www.cafc.uscourts.gov/home/case-information/opinions-orders/"

async def scrape_opinions() -> List[Dict]:
    async with httpx.AsyncClient(timeout=30.0, follow_redirects=True) as client:
        response = await client.get(CAFC_URL)
        response.raise_for_status()
    
    soup = BeautifulSoup(response.text, "html.parser")
    opinions = []
    
    table = soup.find("table")
    if not table:
        raise ValueError("Could not find opinions table on CAFC page")
    
    tbody = table.find("tbody")
    rows = tbody.find_all("tr") if tbody else table.find_all("tr")[1:]
    
    for row in rows:
        cells = row.find_all("td")
        if len(cells) < 6:
            continue
        
        release_date = cells[0].get_text(strip=True)
        appeal_no = cells[1].get_text(strip=True)
        origin = cells[2].get_text(strip=True)
        document_type = cells[3].get_text(strip=True)
        
        case_cell = cells[4]
        case_name = case_cell.get_text(strip=True)
        link = case_cell.find("a")
        
        status = cells[5].get_text(strip=True) if len(cells) > 5 else ""
        
        if not link or not link.get("href"):
            continue
        
        pdf_url = link.get("href")
        if pdf_url.startswith("/"):
            pdf_url = "https://www.cafc.uscourts.gov" + pdf_url
        
        if status != "Precedential" or document_type != "OPINION":
            continue
        
        opinions.append({
            "case_name": case_name,
            "appeal_no": appeal_no,
            "release_date": release_date,
            "origin": origin,
            "document_type": document_type,
            "status": status,
            "pdf_url": pdf_url,
        })
    
    return opinions
